Pin messages to a tribe feature

MathRoda opened the pull request:
https://user-images.githubusercontent.com/92212925/200430451-f600b544-f3a7-4bc5-9813-cffafd14268e.mp4

#498

tomastiminskas commented:
@MathRoda Could you please merge develop branch into the branch you are working on? I just merged lots of changes into develop so it would be good for you to update your branch with those changes

kevkevinpal commented:
running the code on my phone its looking good as an admin,

right now it seems we're not pinning the message as a tribe member. We'll need to parse the chats and then pin the last message to have a value in the `pin` field

haven't reviewed the actual code yet but looking good so far.

We might want to open this PR to another branch other than `master` and `develop` since we still need to do some work for the tribe member to see the pinned message

nice work @MathRoda

tomastiminskas commented:
@kevkevinpal this PR it's done to be merged into stakwork:feat/pinnedMessages so we are good. 

For the tribe member side, we need the following changes on relay: 
- `pin` field on chat to get the uuid of the pined message
- the history received from admin when joining a tribe (which usually send the last 40 messages or so) should also send the pinned message even if it is not one of the last 40 messages. So tribe member have it on device db to show it as pinned on the chat view

Not sure if there's something done regarding that

kevkevinpal commented on sphinx/screens/chats/chat-tribe/chat-tribe/src/main/java/chat/sphinx/chat_tribe/ui/ChatTribeViewModel.kt:
```diff
@@ -271,6 +310,26 @@ internal class ChatTribeViewModel @Inject constructor(
         }
+
+        pinJob?.cancel()
+        pinJob = viewModelScope.launch(mainImmediate) {
+            chatRepository.getChatById(chatId).firstOrNull()?.let { chat ->
+                chat.pinedMessage?.let { messageId ->
+                    messageRepository.getMessageById(messageId).firstOrNull()?.let { message ->
+                        pinedMessageHeaderViewState.updateViewState(
+                            PinedMessageHeaderViewState.PinedMessageHeader(
+                                message
+                            )
+                        )
+
+                    }
+                } ?: run {
+                    pinedMessageHeaderViewState.updateViewState(
+                        PinedMessageHeaderViewState.Idle
+                    )
+                }
+            }
+        }
```
is this where we're parsing the data to see which message to pin?

Just trying to understand the code a bit

MathRoda replied:
yes it is, it should work fine on the tribe member side
i implemented this to retrieve the pin message both for owner& member 
when ever the chat opens and the pin field is not empty it shows the pin message

tomastiminskas replied:
This is where it searches for the Message db record of the pinned message and it shows the bar at the top if chat has a pinned message and that message is in the db.

kevkevinpal commented:
> 

yes this should be done on relay already we just need to implement the actual use of it on the mobile app itself. @MathRoda did you implement the part for the tribe member yet?

tomastiminskas commented:
I don't think there's much to do on the app for the tribe member. If relay sends the pinned message uuid to the app on the chat object, then the changes to search for the message on db and show it on the top bar are done. 

@MathRoda can you confirm if I'm right?

MathRoda commented:
> @MathRoda Could you please merge develop branch into the branch you are working on? I just merged lots of changes into develop so it would be good for you to update your branch with those changes



> I don't think there's much to do on the app for the tribe member. If relay sends the pinned message uuid to the app on the chat object, then the changes to search for the message on db and show it on the top bar are done.
> 
> @MathRoda can you confirm if I'm right?

yes think so too, when owner pin message it updates relay and relay updates chat object in db 
then you go from there, you could control the visibility of the pin message for both member&owner

MathRoda commented:
> > 
> 
> yes this should be done on relay already we just need to implement the actual use of it on the mobile app itself. @MathRoda did you implement the part for the tribe member yet?

yes i did, but it need to be tested from the tribe member view 
my pc couldn't handle running 2 emu with local dev setup

MathRoda commented:
> running the code on my phone its looking good as an admin,
> 
> right now it seems we're not pinning the message as a tribe member. We'll need to parse the chats and then pin the last message to have a value in the `pin` field
> 
> haven't reviewed the actual code yet but looking good so far.
> 
> We might want to open this PR to another branch other than `master` and `develop` since we still need to do some work for the tribe member to see the pinned message
> 
> nice work @MathRoda

in requirements it says that only tribe owners should have the ability to pin messages

kevkevinpal commented:
> 

ya that portion is fine tribe owners are only able to pin messages 

just when testing as a tribe member I don't see the pinned message the owner pinned showing up, but I think you mentioned it was difficult for you to test that scenario

tomastiminskas commented:
@MathRoda I was briefly reviewing your code and I noticed a few things that needs to be fixed:

- You added a column on chat table (`pin_message INTEGER as MessageId`). If you check in relay, the `pin` column is a `String` so I understand the value that needs to be set there is not a `MessageId` but a `MessageUUID`. That makes sense since MessageUUID is shared through all the members of a tribe. @kevkevinpal or @Evanfeenstra can you confirm this please?
- When you add a new column to a table you need to create a migration. It's basically a file with a number as name where you put the ALTER TABLE query. Then the column needs to be added in the CREATE TABLE as well but at the end of the columns list (you added it before the last one). That way you can ensure it will work for users that are already using the app. Check for file `9.sqm`, `10.sqm` or any number before those for migration examples.

MathRoda commented:
@tomastiminskas 
if `MessageUUID` is shared across all members and `MessageId ` is not, then this would totally make sense 
and after changes pin message should appear to the tribe members 

i'll do the changes and let you know, but i'll not be able to test it on the tribe member view i can't run 2 smu with docker on my pc

MathRoda commented:
@tomastiminskas @kevkevinpal 
i added the required changes, and tested it from a tribe owner view it works fine 
can someone test it as a tribe member and confirm pinned message shows up please.

kevkevinpal commented:
hmm I'm still not seeing the pinned message show up as the other tribe member :/

MathRoda commented:
@kevkevinpal i must test it on my own, Let's figure out a way so i can test it as tribe member

tomastiminskas commented:
@kevkevinpal @MathRoda I can help testing this on both admin and member side but I don't think it will be on Monday. Or later today. Will keep you posted

MathRoda commented:
@tomastiminskas thank you for the detailed review,
 conflicts resolved, for the other fixes i'll be looking at those later today if i need it any help sure i'll contact you.
@kevkevinpal can you please add new tasks for :

> The pinned message top bar (that shows below the chat header) is breaking the long press message menu causing the message bubble to show out of place when showing menu. 

> When a message is pinned, the message menu should show unpin option instead os pin, but it's not. Check the screenshot added on last item

MathRoda commented:
@tomastiminskas 
fixed all issues mentioned above : 

- resolved merging conflicts 
- fixed menu pop-up incorrect position when pin header is visible by adding the height measurement to the header calculation in `SelectedMessage.instantiate`  also in `MessageListAdapter` added a logic that sets the value based on the pin header visibility
 note: after pining the message you need to leave the chat and comeback to update the view visibility and get the right calculations, this issue should only be visible for tribe owner 
 soluion : i think i could pass the view visibility to viewstate container to be updated immediately but i need thoughts on this approach @tomastiminskas @kevkevinpal 
![Screenshot_20221116_220019](https://user-images.githubusercontent.com/92212925/202283309-12e6c6fe-fbc0-41c2-b171-06a3f3714d37.png)

- on pined message bubble long click, it shows unpin option rather that pin option 

![Screenshot_20221116_205524](https://user-images.githubusercontent.com/92212925/202285181-0b6c9b07-1098-4f32-8de0-86294bfb7142.png)

- not showing pined message details on pinned bottom pop-up, it only happens because the viewstate don't get updated immediately you need to leave the chat and comeback, solved by passing the message object to viewstate updater after clicking `pinMessage` method in `ViewModel` 
![Screenshot_20221116_205451](https://user-images.githubusercontent.com/92212925/202285830-9515444a-3f9e-4799-8267-154df151c6d5.png)

kevkevinpal commented:
finally got a chance to take a look at the code from relay side but are we able to use this function to get the pinned msg

https://github.com/stakwork/sphinx-kotlin/blob/8972a73ce49a46ea8400e7ff5ca9960562fa2225/sphinx/application/network/features/queries/feature-network-query-chat/src/main/java/chat/sphinx/feature_network_query_chat/NetworkQueryChatImpl.kt#L73

seems like we're updating locally on relay and then publishing the pinned message to tribes server, I think other users of a tribe would need to query the tribes server to get the new pinned msg.

Let me know if that makes sense, this would make it work with how relay is now. But otherwise there will need to be a change to send the pinned msg directly to the chat members

MathRoda commented:
@tomastiminskas @kevkevinpal I'm back 🚀
Pinned message not showing up for other tribe members is now fixed
also : 
- merged master into the current branch and resolved the conflicts 
- fixed and issue for the view not included in ChatContact/ ChatGroup as pinned message header is part of MessageListAdapter 

|   |   |   |   |   | 
|---|---|---|---|---|
|![member_view_1](https://user-images.githubusercontent.com/92212925/230514570-9e4dcdfa-bb53-4c54-8ca9-13e852a31df9.png)|![Screenshot_20230407_015913_chat sphinx debug](https://user-images.githubusercontent.com/92212925/230514644-5db3f35c-566e-46cb-a9d3-c7ab72dda8fd.jpg)|![member_view_2](https://user-images.githubusercontent.com/92212925/230514756-6fc4968a-2a67-4fde-8f3b-1499c28d506c.png)|![Screenshot_20230407_015919_chat sphinx debug](https://user-images.githubusercontent.com/92212925/230514781-3d0633c9-ee8c-4465-8cca-abb3df097eb7.jpg)

tomastiminskas commented:
Thanks @MathRoda I will take a look and test it as soon as possible (I guess it will be on Monday). Do you think we can make "Pinned Message" bottom view adapt to its content? It looks like there's a big empty space below the pinned message and that doesn't look that well. Just a minor UI adjustment, but would be good to solve it while you are working on this.

MathRoda commented:
@tomastiminskas change has been made test it and let me know.
